File: zeroday-server/src/main/kotlin/com/zeroday/handler/handlers/SubscriptionHandler.kt
package com.zeroday.handler.handlers

import com.zeroday.handler.HandlerContext
import com.zeroday.handler.HandlerResult
import com.zeroday.handler.MessageHandler
import com.zeroday.protocol.ErrorCategory
import com.zeroday.protocol.RequestTypes
import com.zeroday.protocol.ResponseTypes
import com.zeroday.protocol.ServerError
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonPrimitive

/**
 * Manages per-connection channel subscriptions.
 *
 * By default every connection subscribes to `player` and `notifications`;
 * all other channels are opt-in so the client doesn't receive frames it
 * has no use for. This is the single most effective bandwidth optimisation
 * in the server: an idle player who never touches inventory or tasks
 * doesn't receive inventory-sync frames every tick.
 *
 * The client sends:
 *   `{"type":"subscribe","payload":{"channels":["inventory","tasks"]}}`
 *   `{"type":"unsubscribe","payload":{"channels":["tasks"]}}`
 */
class SubscriptionHandler : MessageHandler {
    override val handledTypes: Set<String> = setOf(
        RequestTypes.SUBSCRIBE,
        RequestTypes.UNSUBSCRIBE
    )
    override val requiresAuth: Boolean = true

    override suspend fun handle(ctx: HandlerContext, type: String, payload: JsonObject): HandlerResult {
        val rawChannels = payload["channels"]
        val channelNames: List<String> = when (rawChannels) {
            is JsonArray -> rawChannels.mapNotNull {
                (it as? JsonPrimitive)?.takeIf { p -> p.isString }?.content
            }
            else -> return HandlerResult.Err(
                ServerError.fromMessage("'channels' must be an array of channel names", ErrorCategory.VALIDATION)
            )
        }
        if (channelNames.isEmpty()) {
            return HandlerResult.Err(
                ServerError.fromMessage("Empty channel list", ErrorCategory.VALIDATION)
            )
        }
        return when (type) {
            RequestTypes.SUBSCRIBE -> handleSubscribe(ctx, channelNames)
            RequestTypes.UNSUBSCRIBE -> handleUnsubscribe(ctx, channelNames)
            else -> HandlerResult.Err(
                ServerError.fromMessage("Unsupported subscription operation", ErrorCategory.VALIDATION)
            )
        }
    }

    private fun handleSubscribe(ctx: HandlerContext, channels: List<String>): HandlerResult {
        val conn = ctx.connection
        val added = mutableListOf<String>()
        val invalid = mutableListOf<String>()
        for (ch in channels) {
            if (ch in conn.subscribedChannels) continue
            if (ch !in com.zeroday.handler.ConnectionInfo.ALL_CHANNELS) {
                invalid += ch
                continue
            }
            conn.subscribedChannels.add(ch)
            added += ch
        }
        return HandlerResult.Ok(
            type = ResponseTypes.SUBSCRIBED,
            payload = mapOf<String, Any?>(
                "subscribed" to added,
                "invalid" to invalid.ifEmpty { null },
                "active" to conn.subscribedChannels.toList()
            )
        )
    }

    private fun handleUnsubscribe(ctx: HandlerContext, channels: List<String>): HandlerResult {
        val conn = ctx.connection
        val removed = mutableListOf<String>()
        val protected = mutableListOf<String>()
        for (ch in channels) {
            if (ch in com.zeroday.handler.ConnectionInfo.DEFAULT_CHANNELS) {
                protected += ch
                continue
            }
            if (conn.subscribedChannels.remove(ch)) removed += ch
        }
        return HandlerResult.Ok(
            type = ResponseTypes.UNSUBSCRIBED,
            payload = mapOf<String, Any?>(
                "unsubscribed" to removed,
                "protected" to protected.ifEmpty { null },
                "active" to conn.subscribedChannels.toList()
            )
        )
    }
}

File: zero-server-csharp/src/ZeroDayMMO.Server/Handlers/SubscriptionHandler.cs
using System.Text.Json;
using ZeroDayMMO.Shared;

namespace ZeroDayMMO.Server.Handlers;

public class SubscriptionHandler : IHandler
{
    private static readonly HashSet<string> AllChannels = new(StringComparer.OrdinalIgnoreCase)
    {
        "player", "notifications", "inventory", "tasks",
        "world_events", "storyline", "chat", "system"
    };

    private static readonly HashSet<string> DefaultChannels = new(StringComparer.OrdinalIgnoreCase)
    {
        "player", "notifications"
    };

    private readonly Dictionary<string, HashSet<string>> _subscriptions = new();

    public string MessageType => "subscription";

    public Task<IActionResult> HandleAsync(string connectionId, JsonElement? payload, string? requestId)
    {
        if (payload is null)
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "Missing payload" }));

        if (!payload.Value.TryGetProperty("channels", out var channelsElement) || channelsElement.ValueKind != JsonValueKind.Array)
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "'channels' must be an array of channel names" }));

        var channelNames = new List<string>();
        foreach (var item in channelsElement.EnumerateArray())
        {
            if (item.ValueKind == JsonValueKind.String)
                channelNames.Add(item.GetString()!);
        }

        if (channelNames.Count == 0)
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "Empty channel list" }));

        var action = payload.Value.Str("action");
        return action switch
        {
            "subscribe" => HandleSubscribe(connectionId, channelNames),
            "unsubscribe" => HandleUnsubscribe(connectionId, channelNames),
            _ => Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = $"Unknown subscription action: {action}" }))
        };
    }

    private Task<IActionResult> HandleSubscribe(string connectionId, List<string> channels)
    {
        if (!_subscriptions.ContainsKey(connectionId))
            _subscriptions[connectionId] = new HashSet<string>(DefaultChannels, StringComparer.OrdinalIgnoreCase);

        var subs = _subscriptions[connectionId];
        var added = new List<string>();
        var invalid = new List<string>();

        foreach (var ch in channels)
        {
            if (subs.Contains(ch))
                continue;
            if (!AllChannels.Contains(ch))
            {
                invalid.Add(ch);
                continue;
            }
            subs.Add(ch);
            added.Add(ch);
        }

        return Task.FromResult<IActionResult>(
            new MessageResult("subscribed", connectionId, new
            {
                subscribed = added,
                invalid = invalid.Count > 0 ? invalid : null,
                active = subs.ToList()
            }));
    }

    private Task<IActionResult> HandleUnsubscribe(string connectionId, List<string> channels)
    {
        if (!_subscriptions.TryGetValue(connectionId, out var subs))
            return Task.FromResult<IActionResult>(new MessageResult("unsubscribed", connectionId, new
            {
                unsubscribed = new List<string>(),
                @protected = (object?)null,
                active = new List<string>()
            }));

        var removed = new List<string>();
        var protectedChannels = new List<string>();

        foreach (var ch in channels)
        {
            if (DefaultChannels.Contains(ch))
            {
                protectedChannels.Add(ch);
                continue;
            }
            if (subs.Remove(ch))
                removed.Add(ch);
        }

        return Task.FromResult<IActionResult>(
            new MessageResult("unsubscribed", connectionId, new
            {
                unsubscribed = removed,
                @protected = protectedChannels.Count > 0 ? protectedChannels : (object?)null,
                active = subs.ToList()
            }));
    }
}
